### PR TITLE
DATAJPA-1303. Added support ignoring case for finding collections

### PR DESCRIPTION
**Pull request by krupt**
- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAJPA).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@krupt Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-jpa&pullRequestId=261)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-jpa&pullRequestId=261) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@krupt Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-jpa&pullRequestId=261)!

File: src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java [hunk 1]
```diff
@@ -185,19 +188,21 @@ private ParameterMetadataProvider(CriteriaBuilder builder, @Nullable Iterator<Ob
...
 
+		private final Part part;
 		private final Type type;
 		private final ParameterExpression<T> expression;
 		private final PersistenceProvider persistenceProvider;
 
 		/**
 		 * Creates a new {@link ParameterMetadata}.
 		 */
-		public ParameterMetadata(ParameterExpression<T> expression, Type type, @Nullable Object value,
```
Comment (olivergierke) [inline src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java]: Can you clarify why this change is needed? I don't feel comfortable with making `ParameterMetadata` aware of anything ignore-case related. Shouldn't the handling of that be applied in the code eventually calling `….prepare(…)`?

File: src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java [hunk 2]
```diff
@@ -232,19 +237,19 @@ public Object prepare(Object value) {
 
 				switch (type) {
 					case STARTING_WITH:
-						return String.format("%s%%", value.toString());
```
Comment (olivergierke) [inline src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java]: Please avoid changes unrelated to the actual improvement we're trying to fix.

File: src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java [hunk 3]
```diff
@@ -232,19 +237,19 @@ public Object prepare(Object value) {
...
 					case CONTAINING:
 					case NOT_CONTAINING:
-						return String.format("%%%s%%", value.toString());
+						return "%" + value + "%";
 					default:
 						return value;
 				}
 			}
 
-			return Collection.class.isAssignableFrom(expressionType) //
```
Comment (olivergierke) [inline src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java]: Please avoid changes unrelated to the actual improvement we're trying to fix.

File: src/test/java/org/springframework/data/jpa/repository/UserRepositoryTests.java [hunk 4]
```diff
@@ -1175,6 +1175,48 @@ public void findByElementCollectionAttribute() {
 		assertThat(result).containsOnly(firstUser, secondUser);
 	}
 
+	@Test
```
Comment (olivergierke) [inline src/test/java/org/springframework/data/jpa/repository/UserRepositoryTests.java]: Please add references to the JIRA issue like the other test methods do. Also, please add new test methods *last*, not in the middle of others.

File: src/test/java/org/springframework/data/jpa/repository/sample/UserRepository.java [hunk 5]
```diff
@@ -290,6 +290,12 @@
 	// DATAJPA-496
 	List<User> findByAttributesIn(Set<String> attributes);
 
+	List<User> findByAttributesIgnoreCaseIn(Collection<String> attributes);
```
Comment (olivergierke) [inline src/test/java/org/springframework/data/jpa/repository/sample/UserRepository.java]: Please move those methods to the bottom of the file and mark them with the JIRA ticket because of which they were introduced.

File: src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java (hunk 2 above)
Comment (krupt) [reply src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java]: Should I create another issue in JIRA to fix creation of unnecessary objects?

Comment (olivergierke) [reply src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java]: Unless you have detailed profiling numbers showing that that change has significant impact on the performance I don't think it's worth the effort.

File: src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java (hunk 1 above)
Comment (krupt) [reply src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java]: I need more time to study this proposal

Comment (krupt) [reply src/main/java/org/springframework/data/jpa/repository/query/ParameterMetadataProvider.java]: I can't find better solution. Method .prepare calling only from org.springframework.data.jpa.repository.query.QueryParameterSetterFactory.CriteriaQueryParameterSetterFactory#CriteriaQueryParameterSetterFactory, but there have not place for this. In ParameterMetadata we prepare another string conversion and can do this conversion. I think that I should change ignoreCaseType field to boolean

**Comment by schauder** [comment]
Thanks, that is squashed, polished and merged.
